File: packages/engine/src/planning/validate.ts
import type {
  CanonicalItem,
  Household,
  HouseholdRecipeState,
  Recipe,
} from "../schemas/domain.js";
import type { PlanProposal, Violation } from "../schemas/plan.js";
import { runAllergenBackstop } from "../allergen/backstop.js";
import { orphanedPerishables } from "./perishables.js";
import { validateVariety, type VarietyRules, DEFAULT_VARIETY_RULES } from "./variety.js";

export interface ValidatePlanInput {
  proposal: PlanProposal;
  /** Candidate recipes the proposal was allowed to draw from. */
  candidates: ReadonlyMap<string, Recipe>;
  states: ReadonlyMap<string, HouseholdRecipeState>;
  items: ReadonlyMap<string, CanonicalItem>;
  household: Household;
  varietyRules?: VarietyRules;
}

export interface ValidationResult {
  ok: boolean;
  errors: Violation[];
  warnings: Violation[];
}

/**
 * Deterministic validation of a Claude plan proposal. Errors (allergy hard
 * fails above all) mean the plan must not be shown or saved as-is: re-prompt
 * with the violation list, and if the violation persists, drop/replace the
 * offending entry from the pre-filtered safe pool in code.
 */
export function validatePlan(input: ValidatePlanInput): ValidationResult {
  const violations: Violation[] = [];

  input.proposal.entries.forEach((entry, entryIndex) => {
    const recipe = input.candidates.get(entry.recipeId);
    if (!recipe) return; // reported as unknown_recipe by validateVariety
    violations.push(
      ...runAllergenBackstop({
        entry,
        entryIndex,
        recipe,
        people: input.household.people,
        restrictions: input.household.restrictions,
        items: input.items,
      }),
    );
  });

  violations.push(
    ...validateVariety(
      input.proposal.entries,
      input.candidates,
      input.states,
      input.varietyRules ?? DEFAULT_VARIETY_RULES,
    ),
  );

  const plannedRecipes = input.proposal.entries
    .map((e) => input.candidates.get(e.recipeId))
    .filter((r): r is Recipe => r !== undefined);
  violations.push(...orphanedPerishables(plannedRecipes, input.items));

  const errors = violations.filter((v) => v.severity === "error");
  const warnings = violations.filter((v) => v.severity === "warning");
  return { ok: errors.length === 0, errors, warnings };
}

File: service/mealplanner/planning/validate.py
"""Plan validation orchestrator — port of packages/engine/src/planning/validate.ts.

Errors (allergy hard fails above all) mean the plan must not be emailed or
saved as-is: re-prompt with the violation list, and if the violation persists,
drop/replace the offending entry deterministically.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping

from ..allergen.backstop import run_allergen_backstop
from ..schemas.domain import CanonicalItem, Household, HouseholdRecipeState, Recipe
from ..schemas.plan import PlanProposal, Violation
from .perishables import orphaned_perishables
from .variety import DEFAULT_VARIETY_RULES, VarietyRules, validate_variety


@dataclass(frozen=True)
class ValidationResult:
    ok: bool
    errors: list[Violation]
    warnings: list[Violation]


def validate_plan(
    proposal: PlanProposal,
    candidates: Mapping[str, Recipe],
    states: Mapping[str, HouseholdRecipeState],
    items: Mapping[str, CanonicalItem],
    household: Household,
    variety_rules: VarietyRules = DEFAULT_VARIETY_RULES,
) -> ValidationResult:
    violations: list[Violation] = []

    for entry_index, entry in enumerate(proposal.entries):
        recipe = candidates.get(entry.recipe_id)
        if recipe is None:
            continue  # reported as unknown_recipe by validate_variety
        violations.extend(
            run_allergen_backstop(
                entry=entry,
                entry_index=entry_index,
                recipe=recipe,
                people=household.people,
                restrictions=household.restrictions,
                items=items,
            )
        )

    violations.extend(
        validate_variety(proposal.entries, candidates, states, variety_rules)
    )

    planned = [
        r for e in proposal.entries if (r := candidates.get(e.recipe_id)) is not None
    ]
    violations.extend(orphaned_perishables(planned, items))

    errors = [v for v in violations if v.severity == "error"]
    warnings = [v for v in violations if v.severity == "warning"]
    return ValidationResult(ok=not errors, errors=errors, warnings=warnings)
